nep-20.mediawiki

```diff
@@ -0,0 +1,419 @@
+    // Properties
+    compatibility: string[];
+    extra: any;
+    name: string;
+    network: Network;
+    supportedNetworks: Network[];
+    version: string;
+    website: string;
+
+    // Events
+    on(event: EventName, listener: () => void): void;
+    removeListener(event: EventName, listener: () => void): void;
+
+    // Methods
+    authenticate(payload: AuthenticationChallengePayload): Promise<AuthenticationResponsePayload>;
+    call(invocation: InvocationArguments): Promise<InvocationResult>;
+    call(hash: UInt160, operation: string, ...args: Argument[]): Promise<InvocationResult>;
+    getAccounts(): Promise<Account[]>;
+    getApplicationLog(txid: UInt256): Promise<ApplicationLog>;
+    getBalance(asset: UInt160, account?: UInt160): Promise<Integer>;
+    getBlock(hash: UInt256): Promise<Block>;
+    getBlock(index: number): Promise<Block>;
+    getBlockCount(): Promise<number>;
+    getStorage(hash: UInt160, key: Base64Encoded): Promise<Base64Encoded>;
+    getTokenInfo(hash: UInt160): Promise<Token>;
+    getTransaction(txid: UInt256): Promise<Transaction>;
+    invoke(invocations: InvocationArguments[], signers?: Signer[], suggestedSystemFee?: Integer): Promise<UInt256>;
+    makeTransaction(invocations: InvocationArguments[], signers?: Signer[], suggestedSystemFee?: Integer): Promise<ContractParametersContext>;
+    relay(context: ContractParametersContext): Promise<UInt256>;
+    send(asset: UInt160, amount: Integer, to: UInt160, from?: UInt160): Promise<UInt256>;
+    sign(context: ContractParametersContext): Promise<ContractParametersContext>;
+}
+</pre>
+
+===Errors===
+
+<pre>
+export type ErrorType =
+    "UNSUPPORTED" |         // The requested feature or operation is not supported.
+    "INVALID" |             // The input data is in an invalid format.
+    "NOTFOUND" |            // The requested data doesn't exist.
+    "FAILED" |              // The contract execution failed.
+    "TIMEOUT" |             // The requested operation was cancelled due to timeout.
+    "CANCELED" |            // The requested operation was cancelled by the user.
+    "INSUFFICIENT_FUNDS" |  // The requested operation failed due to insufficient balance.
+    "RPC_ERROR" |           // An exception was thrown by the RPC server.
+    "UNKNOWN";              // An unknown error has occurred.
+
+export interface Error {
+    type: ErrorType;
+    message: string;
+    detail?: any;
+}
+
+export interface FailedError extends Error {
+    type: "FAILED";
+    detail: InvocationResult;
+}
+</pre>
+
+===How to obtain an instance of IDapiProvider interface?===
+
+In order for dapps to have a unified method of obtaining <code>IDapiProvider</code> instances, all providers must trigger the <code>Neo.DapiProvider.ready</code> event of the <code>window</code> object when they are ready.
+
+The front end of the dapp can listen to the event and obtain the <code>IDapiProvider</code> instance from the <code>detail</code> property of the event.

```

Comment:
This is fine for web based dapp clients, but what about mobile or desktop clients?

Reply:
I think it is difficult for mobile clients and desktop clients to have a unified method to obtain `IDapiProvider` objects.

Reply:
It doesn't have to be a unified method, but I'd expect there to be *some* method for mobile and desktop apps to access a DAPI provider

Reply:
Do you have any ideas?

Reply:
WalletConnect 2.0 handles this using a "connection code": It could open the wallet via Deeplinking or show the code as an QRCode image, and then, on the Wallet the user can confirm the connection. So, for this to work, we would need a new event, so the dapp can handle this UI interaction to show the code.